### PR TITLE
eth, xeth: fix GasPriceOracle goroutine leak

### LINKED ISSUE #1317: geth 0.9.31 fatal error: runtime: out of memory
I looked through the both open and closed issues and did not find this.

geth was just running and it crashed:

http://paste.ubuntu.com/11758673/

### PR DESCRIPTION
**Pull request by fjl**
XEth.gpo was being initialized as needed. WithState copies the XEth
struct including the gpo field. If gpo was nil at the time of the copy
and Call or Transact were invoked on it, an additional GPO listenLoop
would be spawned.

Move the lazy initialization to GasPriceOracle instead so the same GPO
instance is shared among all created XEths.

Fixes #1317
Might help with #1930

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Mon Oct 26 20:49:45 UTC 2015_

**Comment by bas-vk** [comment]
@fjl, question. Each time `GasPriceOracle.SuggestPrice` is called the `init` method is internally called which executes some initialization code only the first time. Is there a reason why this `init` method isn't called in `NewGasPriceOracle`?

**Comment by fjl** [comment]
There also seems to be an issue (not fixed here) with the blocks map in the GPO.
The map remembers the price for past blocks but entries are never deleted.

**Comment by fjl** [comment]
@bas-vk This preserves the lazy initialization which was present in xeth before. The GPO event listener
is not required unless the user is sending transactions or calls. If neither of these happen,
the listener is not started.

File: eth/gasprice.go [hunk 1]
```diff
@@ -23,49 +23,65 @@ import (
...
+	// for testing
+	gpoDefaultBaseCorrectionFactor = 110
+	gpoDefaultMinGasPrice          = 10000000000000
+)
 
 type blockPriceInfo struct {
 	baseGasPrice *big.Int
 }
 
+// GasPriceOracle provides recommends gas prices based on the content
```
Comment (Gustav-Simonsson) [inline eth/gasprice.go]: can drop either "provides" or "recommends"

File: eth/gasprice.go [hunk 2]
```diff
@@ -176,28 +193,26 @@ func (self *GasPriceOracle) lowestPrice(block *types.Block) *big.Int {
 	return minPrice
 }
 
+// SuggestPrice returns the recommended gas price.
+// The return value may not be modified by the caller.
```
Comment (Gustav-Simonsson) [inline eth/gasprice.go]: Could this restriction be lifted by returning a new big.Int copy of the recommended price? It seems like a odd restriction for a exported function unless it's expected to be called very often where allocations could be a performance / memory bottleneck.

**Comment by fjl** [comment]
@Gustav-Simonsson PTAL

**Comment by bas-vk** [comment]
:+1:

**Comment by zsfelfoldi** [comment]
:+1:

**Pull request merged**